File: api-server/services/query/operator_catalog_test.go
package query

import (
	"strings"
	"testing"
)

// TestOperatorCatalogCoversAllTokens fails if a BinaryWhereClauseType constant
// is declared in entity_query.go but has no entry in OperatorCatalog. This is
// one half of the catalog drift guard — the other half lives in
// api-server/services/observability/operator_catalog_coverage_test.go and
// checks every *Source.GetSupportedOperators() token.
func TestOperatorCatalogCoversAllTokens(t *testing.T) {
	declared := []BinaryWhereClauseType{
		Eq, Nq, Lt, Lte, Gt, Gte, In, NotIn,
		Like, ILike, NLike,
		Contains, IContains, NIContains,
		Regex, NRegex,
		HasKey, IsNull, Between,
		EqF, NqF, LtF, LteF, GtF, GteF, LikeF, ILikeF,
	}
	for _, tok := range declared {
		if _, ok := OperatorCatalog[string(tok)]; !ok {
			t.Errorf("BinaryWhereClauseType %q has no OperatorCatalog entry", tok)
		}
	}
}

// TestOperatorCatalogLabelsPopulated asserts every descriptor carries a
// non-empty label for each kind it claims. A "chip" descriptor with no
// chip_label (or "line" with no line_label) will render the raw token in the
// UI dropdown — the failure mode that motivated issue #29227. Catching it
// here blocks the PR before the UX regression ships.
func TestOperatorCatalogLabelsPopulated(t *testing.T) {
	for token, desc := range OperatorCatalog {
		validateCatalogEntry(t, token, desc)
	}
}

func validateCatalogEntry(t *testing.T, token string, desc OperatorDescriptor) {
	t.Helper()
	if token == "" {
		t.Errorf("OperatorCatalog has empty token key for %+v", desc)
	}
	if !strings.HasPrefix(token, "_") {
		t.Errorf("token %q does not start with '_' — all operator tokens must follow the _<op> convention", token)
	}
	if desc.Token != token {
		t.Errorf("descriptor Token %q does not match catalog key %q", desc.Token, token)
	}
	if len(desc.Kinds) == 0 {
		t.Errorf("descriptor for %q has no Kinds — must contain at least one of chip/line", token)
	}
	for _, k := range desc.Kinds {
		validateCatalogKind(t, token, desc, k)
	}
}

func validateCatalogKind(t *testing.T, token string, desc OperatorDescriptor, kind string) {
	t.Helper()
	switch kind {
	case "chip":
		if desc.ChipLabel == "" {
			t.Errorf("descriptor for %q declares kind=chip but has empty ChipLabel", token)
		}
	case "line":
		if desc.LineLabel == "" {
			t.Errorf("descriptor for %q declares kind=line but has empty LineLabel", token)
		}
	default:
		t.Errorf("descriptor for %q has unknown kind %q — allowed: chip, line", token, kind)
	}
}

// TestDescribeOperatorsSkipsUnknown ensures DescribeOperators tolerates
// unknown tokens without panicking — a defensive behavior since the coverage
// test will flag the real problem.
func TestDescribeOperatorsSkipsUnknown(t *testing.T) {
	got := DescribeOperators([]string{string(Eq), "_bogus_token", string(Contains)})
	if len(got) != 2 {
		t.Fatalf("expected 2 descriptors (unknown skipped), got %d", len(got))
	}
	if got[0].Token != string(Eq) || got[1].Token != string(Contains) {
		t.Errorf("unexpected descriptors: %+v", got)
	}
}

File: app/__tests__/components1/k8s/common/operatorCatalog.test.ts
import { getOperatorDisplayLabel, getOperatorsForKind, normalizeLegacyOperator, OperatorDescriptor } from '@components1/k8s/common/operatorCatalog';

const SAMPLE_DESCRIPTORS: OperatorDescriptor[] = [
  { token: '_eq', chip_label: '=', kinds: ['chip'] },
  { token: '_neq', chip_label: '!=', kinds: ['chip'] },
  { token: '_contains', chip_label: 'contains', line_label: 'Line contains', kinds: ['chip', 'line'] },
  { token: '_regex', chip_label: '=~', line_label: 'Line contains regex match', kinds: ['chip', 'line'] },
  { token: '_line_only', line_label: 'Line only', kinds: ['line'] },
];

describe('operatorCatalog', () => {
  describe('getOperatorsForKind', () => {
    it('filters descriptors by chip kind and uses chip_label', () => {
      expect(getOperatorsForKind(SAMPLE_DESCRIPTORS, 'chip')).toEqual([
        { label: '=', value: '_eq' },
        { label: '!=', value: '_neq' },
        { label: 'contains', value: '_contains' },
        { label: '=~', value: '_regex' },
      ]);
    });

    it('filters descriptors by line kind and uses line_label', () => {
      expect(getOperatorsForKind(SAMPLE_DESCRIPTORS, 'line')).toEqual([
        { label: 'Line contains', value: '_contains' },
        { label: 'Line contains regex match', value: '_regex' },
        { label: 'Line only', value: '_line_only' },
      ]);
    });

    it('returns an empty array when descriptors is undefined', () => {
      expect(getOperatorsForKind(undefined, 'chip')).toEqual([]);
      expect(getOperatorsForKind(undefined, 'line')).toEqual([]);
    });

    it('returns an empty array when descriptors is an empty array', () => {
      expect(getOperatorsForKind([], 'chip')).toEqual([]);
      expect(getOperatorsForKind([], 'line')).toEqual([]);
    });

    it('falls back to the token when the requested kind label is missing', () => {
      const descriptors: OperatorDescriptor[] = [{ token: '_no_labels', kinds: ['chip'] }];
      expect(getOperatorsForKind(descriptors, 'chip')).toEqual([{ label: '_no_labels', value: '_no_labels' }]);
    });
  });

  describe('normalizeLegacyOperator', () => {
    it.each([
      ['=', '_eq'],
      ['!=', '_neq'],
      ['<', '_lt'],
      ['<=', '_lte'],
      ['>', '_gt'],
      ['>=', '_gte'],
      ['=~', '_regex'],
      ['!~', '_nregex'],
      ['CONTAINS', '_contains'],
      ['ICONTAINS', '_icontains'],
      ['NOT ICONTAINS', '_nlike'],
      ['LIKE', '_like'],
      ['ILIKE', '_ilike'],
      ['NOT LIKE', '_nlike'],
      ['NOT ILIKE', '_nlike'],
      ['REGEX', '_regex'],
      ['NOT REGEX', '_nregex'],
      ['REGEXP', '_regex'],
      ['IN', '_in'],
      ['NOT IN', '_not_in'],
      ['EXISTS', '_has_key'],
      ['NOT EXISTS', '_is_null'],
      ['BETWEEN', '_between'],
    ])('maps legacy value %s to backend token %s', (legacy, token) => {
      expect(normalizeLegacyOperator(legacy)).toBe(token);
    });

    it('passes backend tokens through unchanged (idempotent)', () => {
      expect(normalizeLegacyOperator('_eq')).toBe('_eq');
      expect(normalizeLegacyOperator('_contains')).toBe('_contains');
    });

    it('passes unknown values through unchanged', () => {
      expect(normalizeLegacyOperator('HOPEFULLY_UNDEFINED')).toBe('HOPEFULLY_UNDEFINED');
    });
  });

  describe('getOperatorDisplayLabel', () => {
    it('resolves a backend token to its chip_label via descriptor lookup', () => {
      expect(getOperatorDisplayLabel('_contains', SAMPLE_DESCRIPTORS)).toBe('contains');
    });

    it('normalizes a legacy value and resolves the chip_label', () => {
      expect(getOperatorDisplayLabel('CONTAINS', SAMPLE_DESCRIPTORS)).toBe('contains');
    });

    it('falls back to the raw op when no descriptor matches', () => {
      expect(getOperatorDisplayLabel('NEW_OPERATOR', SAMPLE_DESCRIPTORS)).toBe('NEW_OPERATOR');
    });

    it('returns an empty string for falsy input', () => {
      expect(getOperatorDisplayLabel('', SAMPLE_DESCRIPTORS)).toBe('');
    });

    it('returns the raw op when descriptors is undefined', () => {
      expect(getOperatorDisplayLabel('_eq', undefined)).toBe('_eq');
    });

    it('returns the raw op when descriptors is an empty array', () => {
      expect(getOperatorDisplayLabel('_eq', [])).toBe('_eq');
    });
  });
});
